### PR TITLE
Added support for inputs and sample rates(macOS)

### PR DESCRIPTION
**Pull request by c0deSpace**
Added analog inputs and sample rates.
The buffer size is reduced, otherwise the recorded sound is cracked. Playback does not crack with increased and decreased buffer size. Tested on my E-MU 1616m. According to reports, Audigy (sb0092), SB Live (sb0102) also work well.

### CODE REVIEW

**Review by kxproject — CHANGES_REQUESTED** [review]
-- duplicate, added per-line comments in a separate review --

**Review by kxproject — DISMISSED** [review]

**Review by kxproject — CHANGES_REQUESTED** [review]
-- duplicate, added per-line comments in a separate review --

**Review by kxproject — CHANGES_REQUESTED** [review]
-- duplicate, added per-line comments in a separate review --

File: macosx/kext/AudioDevice.cpp
Comment (kxproject) [inline macosx/kext/AudioDevice.cpp:467]: There's one kx_lock_release() above. Should remove the first one and keep this one.

File: macosx/kext/AudioDevice.cpp
Comment (kxproject) [inline macosx/kext/AudioDevice.cpp:572]: kIOAudioControlChannelNameRight is missing, kIOAudioControlChannelNameAll is not required.

File: macosx/kext/AudioDevice.cpp
Comment (kxproject) [inline macosx/kext/AudioDevice.cpp:837]: Why not to use kIOAudioControlChannelIDDefaultLeft and similar constants here?

File: macosx/kext/AudioEngine.cpp
Comment (kxproject) [inline macosx/kext/AudioEngine.cpp]: The comment above suggests that on Mavericks there's a buffer underrun. I wonder if it is specific to some earlier macOS release or to all future releases as well. Not sure how removing *4 multiplier affects other systems.

File: macosx/kext/AudioEngine.cpp
Comment (kxproject) [inline macosx/kext/AudioEngine.cpp]: 28 was taken from 10k1/10k2 datasheet specs. Where '40' comes from?

**Review by kxproject — CHANGES_REQUESTED** [review]
Cannot merge this pull request unless issues are fixed. This time I added comments to the code.

File: macosx/kext/AudioDevice.cpp
Comment (kxproject) [inline macosx/kext/AudioDevice.cpp]: I suggest "// these are +12dB volumes" comment goes here.

File: macosx/kxac97/Ac97App.mm
Comment (kxproject) [inline macosx/kxac97/Ac97App.mm]: nItem is always '0' and is never assigned. This breaks 'for' cycle below.

File: macosx/kxac97/Ac97App.mm
Comment (kxproject) [inline macosx/kxac97/Ac97App.mm:108]: nItem is always '0', need to rewrite the code to fix the problem.

**Review by kxproject — CHANGES_REQUESTED** [review]
This commit hides the bug, won't fix it.

File: macosx/kext/AudioEngine.cpp
Comment (kxproject) [inline macosx/kext/AudioEngine.cpp:505]: 44.1 is not supported on 10k1 devices without hardware SRC
hardware SRC is much worse than modern software SRC implemented by apple in CoreAudio
I doubt we need to support 44.1 on 10k1 devices
The same applies to 88.2, 96 and 176.4 below

File: macosx/kext/AudioEngine.cpp
Comment (kxproject) [inline macosx/kext/AudioEngine.cpp]: While this adds support for other sampling rates, in reality it produces worse audio quality output.
Hardware SRC on 10k1/10k2 devices was great 15 years ago, but modern software is better :)

File: macosx/kext/AudioEngine.cpp
Comment (kxproject) [inline macosx/kext/AudioEngine.cpp:476]: True 24/32-bit support requires b2b plugins in the DSP + extra handling of audio data.

File: macosx/kext/AudioEngine.cpp
Comment (kxproject) [inline macosx/kext/AudioEngine.cpp:718]: This is useful, yes. 44.1 is natively supported by E-DSP.

For E-DSP to handle 88.2/96 or 176.4/192kHz sampling rates, extra processing of audio data is required.
(And also WCLOCK_2X / 4X should be set).
This is fully implemented in the 2nd Windows driver (kx-1010-driver), but not supported by Windows kX driver and macOS driver unless somebody adds this support.

**Review by kxproject — CHANGES_REQUESTED** [review]
At present, I see no reason in adding hardware SRC-based formats. Since 10k1/10k2 hardware operates internally at 48kHz, built-in SRC produces average results with 44.1/88.2/176.4 sampling rates. A possible workaround is to use p16v/p17 module or E-DSP, but not voicetable[].param.initial_pitch/pitch_target stuff.

File: macosx/kext/AudioDevice.cpp
Comment (c0deSpace) [reply macosx/kext/AudioDevice.cpp:837]: It's clearer than kIOAudioControlChannelIDDefaultRightRear for input.

File: macosx/kext/AudioEngine.cpp
Comment (c0deSpace) [reply macosx/kext/AudioEngine.cpp]: With such a value the recorded sound is normal for me.

File: macosx/kxac97/Ac97App.mm
Comment (c0deSpace) [reply macosx/kxac97/Ac97App.mm:108]: Forgot insert the increment. nItem++;

File: macosx/kext/AudioEngine.cpp
Comment (c0deSpace) [reply macosx/kext/AudioEngine.cpp:505]: I do not understand how this algorithm works (voicetable[].param.initial_pitch/pitch_target)

File: macosx/kext/AudioEngine.cpp
Comment (c0deSpace) [reply macosx/kext/AudioEngine.cpp]: At any time can choose 48kHz.

File: macosx/kext/AudioEngine.cpp
Comment (c0deSpace) [reply macosx/kext/AudioEngine.cpp:718]: I tried (WCLOCK_2X / 4X ) but there is no result yet. Something else is needed.

Я бы предпочел говорить на русском.

File: macosx/kext/AudioEngine.cpp
Comment (c0deSpace) [reply macosx/kext/AudioEngine.cpp]: When size*4, in recorded sound have gaps.

**Comment by c0deSpace** [comment]
I closed the pool request, because my repository did not the contain latest kX code.

**Comment by c0deSpace** [comment]
I closed the pull request, because my repository did not the contain latest kX code.

**Comment by c0deSpace** [comment]
I closed the pull request, because my repository did not contain latest kX code.

File: macosx/kext/AudioDevice.cpp
Comment (kxproject) [reply macosx/kext/AudioDevice.cpp:837]: Ah, yes, ok, I agree. Let's use plain numbers for recording channels.

File: macosx/kext/AudioEngine.cpp
Comment (kxproject) [reply macosx/kext/AudioEngine.cpp:505]: There are two different formulas used for calculating pitch and sampling rate. Pitch is relative to 48000 and is measured in 'cents'.

File: macosx/kext/AudioEngine.cpp
Comment (kxproject) [reply macosx/kext/AudioEngine.cpp:718]: In order to support native 88.2/96 and 176.4/196kHz modes on 1010 hardware, one needs to split audio channel into separate 16-bit 48kHz stereo streams (24/88200 mono stream becomes (2x2)x16x48000 stream, that is, 4 mono 16/48 streams are required) and later 'assemble' them back inside DSP and write into corresponding FX DSP output registers. This is a bit tricky stuff, but it works for 1010 PCI driver. kX driver has no provisions for such operation and requires serious re-write (in other words, it is not a matter of two or four register writes or setting WCLOCK bits).

As for Russian, yes, I understand. But github is an international dev. resource, and most users here are not Russian speakers, unfortunately. Sorry... I hope google translate helps. If you like, you can post in Russian.

**Pull request closed without merging**
